api/agent/drivers/docker/imageCache.go

```diff
@@ -0,0 +1,103 @@
+package docker
+
+import (
+	"container/list"
+	"sync"
+
+	d "github.com/fsouza/go-dockerclient"
+)
+
+type ImageEvictor func(d.APIImages) error
+
+// Cache is an LRU cache, safe for concurrent access.
+type Cache struct {
+	totalSize int64
+	mu        sync.Mutex
+	ll        *list.List
+	cache     map[string]*list.Element
+	maxSize   int64
+	onEvict   ImageEvictor
+}
+
+// *entry is the type stored in each *list.Element.
+type entry struct {
+	key   string
+	value d.APIImages
+}
+
+// New returns a new cache with the provided maximum items.
+func NewLRU(maxSize int, onEvict ImageEvictor) *Cache {
+	return &Cache{
+		ll:    list.New(),
+		cache: make(map[string]*list.Element),
+	}
+}
+
+// Add adds the provided key and value to the cache, evicting
+// an old item if necessary.
+func (c *Cache) Add(key string, value d.APIImages) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Already in cache?
+	if ee, ok := c.cache[key]; ok {
+		c.ll.MoveToFront(ee)
+		ee.Value.(*entry).value = value
```

Comment:
It's a weird edge case, but if value.Size is updated, then we'd need to update totalSize and run LRU logic too.